src-electron/upgrade/upgrade.js

```diff
@@ -0,0 +1,126 @@
+/**
+ *
+ *    Copyright (c) 2020 Silicon Labs
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+const dbApi = require('../db/db-api.js')
+const queryPackage = require('../db/query-package.js')
+const queryAttribute = require('../db/query-attribute.js')
+const queryCluster = require('../db/query-cluster.js')
+const dbEnum = require('../../src-shared/db-enum.js')
+const fs = require('fs')
+const fsp = fs.promises
+
+/**
+ * This file implements upgrade rules which are used to upgrade .zap files and xml files
+ * to be in sync with the spec
+ */
+
+/**
+ * Returns an array of objects containing global attributes that should be forced external.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} attributeId
+ * @returns An array of objects
+ */
+
+async function getForcedExternalStorage(db, attributeId) {
+  let pkgs = await queryPackage.getPackageRefByAttributeId(db, attributeId)
+  let zcl = await queryPackage.getPackageByPackageId(db, pkgs)
+  zcl = zcl.path
+  let obj = await fsp.readFile(zcl)
+  let data = JSON.parse(obj)
+  let byName = data?.attributeAccessInterfaceAttributes
+  let lists = data?.listsUseAttributeAccessInterface
+  let forcedExternal = { byName, lists }
+  return forcedExternal
+}
+
+/**
+ * Returns a flag stating which type of storage option the attribute is categorized to be.
+ *
+ * @export
+ * @param {*} db
+ * @param {*} clusterName
+ * @param {*} clusterRef
+ * @param {*} storagePolicy
+ * @param {*} forcedExternal
+ * @param {*} attributeId
+ * @returns Storage Option
+ */
+
+async function computeStorageNewConfig(
+  db,
+  clusterRef,
+  storagePolicy,
+  forcedExternal,
+  attributeId
+) {
+  let storageOption
+  let attributeName
+  let clusterName
+  clusterName = await queryCluster.selectClusterName(db, clusterRef)
```

Comment:
You don't need to do this. From where you are calling `computeStorageNewConfig` you have access to the cluster name pass that as argument and use directly

Reply:
I'm making this change but for the record, if we want to reuse this function then we will have to use it in a spot in the code that provides all of that data 

For example, if we want to run "computeStorageNewConfig" somewhere else in the code that does not have cluster name then we will not be able to do it. Having the function do all of the work that it requires makes it much more reusable. 

By making this change we are conceding that it will only be used once.